### PR TITLE
extended theme service

### LINKED ISSUE #27628: Email templates cannot be loaded.
### Steps to reproduce

0. Using 10.0.0beta2 tarball do: 
1. Go to settings -> general -> email templates.
2. Try to edit an email template.

### Expected behaviour
You can show the default content and you can edit it yourself.

### Actual behaviour
It fails to load the template.

screen shot 2017-04-12 at 09 39 43

### Server configuration

**Operating system**:
Ubuntu 14.04

**Web server:**
Apache

**Database:**
MySQL

**PHP version:**
7.0

**ownCloud version:** (see ownCloud admin page)
10.0.0beta2 tarball

**Updated from an older ownCloud or fresh install:**
Fresh

**The content of config/config.php:**

```

```

**Are you using external storage, if yes which one:** local/smb/sftp/...
No.

**Are you using encryption:** 
No

**Logs**

```
```
### Client configuration

**browser**
Chrome

### PR DESCRIPTION
**Pull request by phisch**
## Description
This adds the functionality to get a list of all available themes or search for one through the ThemeService.

## Related Issue
Templateditor PR https://github.com/owncloud/templateeditor/pull/52, which fixes https://github.com/owncloud/core/issues/27628, has a hard dependency on this extension

## Motivation and Context
The mail template editor needs to know which themes there are, and i don't want too much theme logic it its repository, this way we can keep it clean.

## How Has This Been Tested?
Partially tested through unittests. Fully tested by hand.
The usual overwrites: css, js, img, templates, translations, mimetype icons all work for legacy and app themes, for subdir and normal deployments.

Tested the following overwrites:

app theme - regular deployment:
- [x] css
- [x] js
- [x] img
- [x] template
- [x] mimetype icons
- [x] translations

app theme - subdir deployment:
- [x] css
- [x] js
- [x] img
- [x] template
- [x] mimetype icons
- [x] translations

legacy theme - regular deployment:
- [x] css
- [x] js
- [x] img
- [x] template
- [x] mimetype icons
- [x] translations

legacy theme - sub dir deployment:
- [x] css
- [x] js
- [x] img
- [x] template
- [x] mimetype icons
- [x] translations

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by PVince81** [comment]
Is this done ? If yes, please set the "3 - To Review" label and summon reviewers.

**Comment by phisch** [comment]
Since the changes are used in an app, should i move this to the public api?

File: lib/private/Theme/ThemeService.php [hunk 1]
```diff
@@ -39,7 +39,12 @@ class ThemeService {
 	 */
 	public function __construct($themeName = '', $defaultThemeDirectory = '') {
 		$this->setDefaultThemeDirectory($defaultThemeDirectory);
-		$this->createTheme($themeName);
+
+		if ($themeName === '' && $this->defaultThemeExists()) {
```
Comment (PVince81) [inline lib/private/Theme/ThemeService.php]: weird, why does Github show `&&` in red now ?

**Comment by PVince81** [comment]
> Since the changes are used in an app, should i move this to the public api?

If some functions need to be called by apps, please make these public.

The code checker would complain when scanning the app if you're using private APIs

**Comment by phisch** [comment]
Ok, going to update.

**Review by DeepDiver1975 — CHANGES_REQUESTED** [review]
public api is required

**Comment by PVince81** [comment]
@phisch any update ?

**Comment by phisch** [comment]
@DeepDiver1975 please give me feedback if this is okay.

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 please give me feedback if this is okay.

OCP shall only hold interfaces - no classes. the idea here is to separate the public interface declaration from its implementation.

**Comment by DeepDiver1975** [comment]
setting back to '2 - developing' because of missing proper public interfaces

**Comment by PVince81** [comment]
Without this it will not be possible to theme the setup page as it was before, which is a regression.

Not sure how bad this is and whether it can wait for 10.0.2 @felixboehm  ...

**Comment by PVince81** [comment]
@phisch or is that only needed for the email template app?

File: lib/private/Server.php [hunk 2]
```diff
@@ -798,6 +798,7 @@ public function __construct($webRoot, \OC\Config $config) {
 		$this->registerService('ThemeService', function ($c) {
 			return new ThemeService($this->getSystemConfig()->getValue('theme'));
 		});
+		$this->registerAlias('IThemeService', 'ThemeService');
```
Comment (DeepDiver1975) [inline lib/private/Server.php]: Try 'OCP\Theme\IThemeService'

**Comment by PVince81** [comment]
@phisch how much work is missing to have this working ?

**Comment by phisch** [comment]
@PVince81 ups, i just forgot to switch to review label again. It should be done.

**Comment by PVince81** [comment]
It's still a huge thing, looks a bit risk to merge now.

**Comment by PVince81** [comment]
moving to 10.0.2

**Comment by PVince81** [comment]
@phisch phinisch this ?

**Comment by phisch** [comment]
@PVince81 this has been complete the whole time, we just did not want to merge because it was too close to a release.
@DeepDiver1975 can you re-review?

File: lib/private/Theme/ThemeService.php (hunk 1 above)
Comment (phisch) [inline lib/private/Theme/ThemeService.php]: Github had a syntax highlighting hickup. (<- just for the history)

**Comment by DeepDiver1975** [comment]
translation have been tested? php as well as js translations? 

I don't see any related change in the code which could make it possible - see https://github.com/owncloud/core/issues/28167#issuecomment-313494564

@phisch

**Comment by phisch** [comment]
I tested the JS translations which could be affected by this. And yes they work as expected.

**Comment by PVince81** [comment]
@phisch can you rebase to refresh CI ?

@DeepDiver1975 so this can be merged ?

**Comment by PVince81** [comment]
rebased

**Comment by PVince81** [comment]
@DeepDiver1975 no feedback from you, assuming you don't have any comment to add

**Pull request merged**